### ISSUE TITLE
Support 401 / 403 / etc. error codes

### ISSUE THREAD

**Issue by dhermes** [GoogleCloudPlatform/go-endpoints#58]
Currently things like

```go
errors.New("No token in the current context.")
```

result in a 400 error every time.

**Comment by adg** [comment]
Could do this with a custom error type:

```
type HTTPError struct {
  Err error
  StatusCode int
}

**Comment by crhym3** [comment]
But those are already defined in https://github.com/GoogleCloudPlatform/go-endpoints/blob/master/endpoints/errors.go

Or is it something else?

**Comment by dhermes** [comment]
DERP. My bad. I just ASSumed they werent' defined since `endpoints.CurrentUser` [doesn't use them][1].

I suppose I should change the issue title to: Document custom errors and use 401 / 403 for relevant auth errors. WDYT?

[1]: https://github.com/GoogleCloudPlatform/go-endpoints/blob/24eb5d4ab4686e08df186593f021bf2d53db40ee/endpoints/auth.go#L543

**Comment by crhym3** [comment]
@dhermes yeah, I think it's a good idea! We should revise all returned errors. I think we were returning those generic errors because the pre-defined ones didn't exist at that point.

**Comment by zboralski** [comment]
We should remove the requirement that the Error messages begin by their equivalent StatusText... Right now if you create a new NotFound error, it will revert to a 400 unless it starts by "Not Found"... I'll post a pr later.

**Comment by sabrinatseng** [comment]
Java 401 error implementation
**OAuth Protecting a method**	https://cloud.google.com/appengine/docs/java/endpoints/getstarted/backend/code_walkthrough#oauth_protecting_a_method
**Live demo**		https://apis-explorer.appspot.com/apis-explorer/?base=https://cloudendpointstic.appspot.com/_ah/api#p/tictactoe/v1/tictactoe.scores.insert		_**Execute without OAuth**_

**Comment by zboralski** [comment]
There is a related issue already... 

https://github.com/GoogleCloudPlatform/go-endpoints/issues/110

Right if error message are not prefixed by the standard http status code's text then are turned into 400.
